src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/ArraySinglePrimitiveRecord.cs

```diff
@@ -0,0 +1,164 @@
+    }
+
+    private protected override void AddValue(object value)
+    {
+        Debug.Fail("This should never happen");
+        throw new InvalidOperationException();
+    }
+
+    internal static IReadOnlyList<T> ReadPrimitiveTypes(BinaryReader reader, int count)
+    {
+        if (typeof(T) == typeof(byte))
+        {
+            return (IReadOnlyList<T>)(object)ReadBytes(reader, count);
+        }
+
+        List<T> values = [];
+        for (int i = 0; i < count; i++)
+        {
+            if (typeof(T) == typeof(bool))
+            {
+                values.Add((T)(object)reader.ReadBoolean());
+            }
+            else if (typeof(T) == typeof(sbyte))
+            {
+                values.Add((T)(object)reader.ReadSByte());
+            }
+            else if (typeof(T) == typeof(char))
+            {
+                values.Add((T)(object)reader.ReadChar());
+            }
+            else if (typeof(T) == typeof(short))
+            {
+                values.Add((T)(object)reader.ReadInt16());
+            }
+            else if (typeof(T) == typeof(ushort))
+            {
+                values.Add((T)(object)reader.ReadUInt16());
+            }
+            else if (typeof(T) == typeof(int))
+            {
+                values.Add((T)(object)reader.ReadInt32());
+            }
+            else if (typeof(T) == typeof(uint))
+            {
+                values.Add((T)(object)reader.ReadUInt32());
+            }
+            else if (typeof(T) == typeof(long))
+            {
+                values.Add((T)(object)reader.ReadInt64());
+            }
+            else if (typeof(T) == typeof(ulong))
+            {
+                values.Add((T)(object)reader.ReadUInt64());
+            }
+            else if (typeof(T) == typeof(float))
+            {
+                values.Add((T)(object)reader.ReadSingle());
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                values.Add((T)(object)reader.ReadDouble());
+            }
+            else if (typeof(T) == typeof(decimal))
+            {
+                values.Add((T)(object)decimal.Parse(reader.ReadString(), CultureInfo.InvariantCulture));
+            }
+            else if (typeof(T) == typeof(DateTime))
+            {
+                values.Add((T)(object)Utils.BinaryReaderExtensions.CreateDateTimeFromData(reader.ReadInt64()));
+            }
+            else
+            {
+                Debug.Assert(typeof(T) == typeof(TimeSpan));
+
+                values.Add((T)(object)new TimeSpan(reader.ReadInt64()));
+            }
+        }
+
+        return values;
+    }
+
+    private static IReadOnlyList<byte> ReadBytes(BinaryReader reader, int count)
+    {
+        // Special casing byte for performance.
+        if (count <= DefaultMaxArrayLength)
+        {
+            return reader.ReadBytes(count);
+        }
+
+        // But only to certain degree, as the input is untrusted.
```

Comment:
What is this actually defending against?

If the payload/stream/whatever says there are ArrayMax elements, this will do:

* Create an array of 64000
* Read 64000 items (let's assume one call)
* Copy 64000 items into the list
* Read 64000 items
* The list now internally doubles its buffer to 128000, copies 64000 to the new buffer.
* Copy 64000 items into the list
* Read/Copy/Read
* The list now doubles again to 256000, copying 128000 items again.
* Copy/Read/Copy/Read
* The list doubles again to 512000, copying 256000 again.
* Copy/Read/Copy/Read/Copy/Read/Copy/Read
* Double to 1024000, copy 512000
* Copy/Read/Copy/Read/Copy/Read/Copy/Read/Copy/Read/Copy/Read/Copy/Read/Copy/Read
* Double to 2048000, copy 1024000
* Copy/Read/etc
* Double 4096000, copy 1024000.  Et al.
* The 15th doubling doubles to 2097152000, copying 1048576000 items
* The 16th doubling should double to 4194304000, which is -100663296, which is turned back to 4194304000, which is bigger than 2146435071, so it gets clamped to 2146435071.  Now copy 2097152000 items into the final buffer, leaving space for the last 49283071 (which will take a further 751 (or more) calls to Read because of the 64000 intermediate (which is actually a 65536 intermediate, because ArrayPool loves powers of 2))

Total work of this approach:
* Rent 1 array.
* Create 16 arrays inside the List, 15 of which are allocated on/in the LOH.
* Copy 4194240000 bytes throughout growing the list.
* Copy 2146435071 bytes into the rented array through calls to Read
* Copy 2146435071 bytes into the returned list

Vs the trivial approach:

* Allocate an array of size 2146435071 via the LOH
* Copy 2146435071 bytes of data into it via Read

So this defense costs 6GB of copying, plus 15 short-lived LOH objects, and a peak memory requirement of around 4GB (the 2097152000 penultimate buffer and the 2146435071 final buffer, which have to both exist concurrently in order to copy one into the other).

The risk of not doing it is that someone can say "I have a MaxArrayLength array... EOF.  Tee he he".  I vaguely believe that the cure is worse than the disease here.

This defense might be OK if you just make your own `IReadOnlyList` type where you can just keep slapping in newly acquired arrays, whether you do it as a bunch of 64k (either 64000 or 65536) arrays and use DivRem for indexing, or a set of ever-increasing arrays and fancier math for indexing.

Reply:
I've addressed this problem in https://github.com/dotnet/runtime/pull/102379/commits/c485e4a1e6633df7b354df19b0d6ea0dfff83802